crates/gen-guest-teavm-java/src/lib.rs

```diff
@@ -0,0 +1,1918 @@
+                    let ty = self.type_name(iface, ty);
+                    format!(
+                        r#"public {ty} get{case_name}() {{
+                               if (this.tag == {tag}) {{
+                                   return ({ty}) this.value;
+                               }} else {{
+                                   throw new RuntimeException("expected {tag}, got " + this.tag);
+                               }}
+                           }}
+                        "#
+                    )
+                })
+            })
+            .collect::<Vec<_>>()
+            .join("\n");
+
+        let tags = variant
+            .cases
+            .iter()
+            .enumerate()
+            .map(|(i, case)| {
+                let tag = case.name.to_shouty_snake_case();
+                format!("public static final {tag_type} {tag} = {i};")
+            })
+            .collect::<Vec<_>>()
+            .join("\n");
+
+        uwrite!(
+            self.src,
+            "
+            public static final class {name} {{
+                public final {tag_type} tag;
+                private final Object value;
+
+                private {name}({tag_type} tag, Object value) {{
+                    this.tag = tag;
+                    this.value = value;
+                }}
+
+                {constructors}
+                {accessors}
+                {tags}
+            }}
+            "
+        );
+    }
+
+    fn type_option(
+        &mut self,
+        iface: &Interface,
+        id: TypeId,
+        name: &str,
+        payload: &Type,
+        docs: &Docs,
+    ) {
+        self.type_variant(
+            iface,
+            id,
+            name,
+            &Variant {
+                cases: vec![
+                    Case {
+                        docs: Docs::default(),
+                        name: "none".into(),
+                        ty: None,
+                    },
+                    Case {
+                        docs: Docs::default(),
+                        name: "some".into(),
+                        ty: Some(*payload),
+                    },
+                ],
+            },
+            docs,
+        )
+    }
+
+    fn type_result(
+        &mut self,
+        iface: &Interface,
+        id: TypeId,
+        name: &str,
+        result: &Result_,
+        docs: &Docs,
+    ) {
+        self.type_variant(
+            iface,
+            id,
+            name,
+            &Variant {
+                cases: vec![
+                    Case {
+                        docs: Docs::default(),
+                        name: "ok".into(),
+                        ty: result.ok,
+                    },
+                    Case {
+                        docs: Docs::default(),
+                        name: "err".into(),
+                        ty: result.err,
+                    },
+                ],
+            },
+            docs,
+        )
+    }
```

Comment:
Options/results are so commonly used and additionally aren't required to be named, so one thing I've done in the Python bindings for example is to create a helper `Result[T, E]` class. Would it be reasonable to use JS generics for something like this to have a `Result<T, E>` or `Option<T>` class?

Or perhaps for `Option<T>` would it perhaps be beneficial to specialize `Option<list<u8>>` for example to a `null` value but `Option<u32>` would generate a class wrapper?